lib/src/modules/download_file.cc

```diff
@@ -113,31 +119,71 @@ namespace Modules {
     return ActionOutput { exit_code, std_out, std_err };
   }
 
+  ActionResponse DownloadFile::failure_response(const ActionRequest& request,
+                                      const fs::path& results_dir,
+                                      const std::string& message)
+  {
+    LOG_ERROR(message);
+    ActionResponse fail_response { ModuleType::Internal, request };
+    processOutputAndUpdateMetadata(fail_response);
+    fail_response.output = write_download_results(results_dir, EXIT_FAILURE, "", message);
+    return fail_response;
+  }
+
   // DownloadFile overrides callAction from the base BoltModule class since there's no need to run
   // any commands with DownloadFile. CallAction will simply download the file and return a result
   // based on if the download succeeded or failed.
   ActionResponse DownloadFile::callAction(const ActionRequest& request)
   {
     auto file_params = request.params();
     auto files = file_params.get<std::vector<lth_jc::JsonContainer>>("files");
-    auto destination = fs::path(files[0].get<std::string>("destination"));
     const fs::path& results_dir = request.resultsDir();
 
     ActionResponse response { ModuleType::Internal, request };
     lth_jc::JsonContainer result;
-    try {
-      Util::downloadFileFromMaster(master_uris_,
-                    file_download_connect_timeout_,
-                    file_download_timeout_,
-                    client_,
-                    module_cache_dir_->createCacheDir(files[0].get<std::string>("sha256")),
-                    destination,
-                    files[0]);
-      response.output = write_download_results(results_dir, EXIT_SUCCESS, "downloaded", "");
-    } catch (Module::ProcessingError& e) {
-      LOG_ERROR("Failed to download {1}; {2}", destination, e.what());
-      response.output = write_download_results(results_dir, EXIT_FAILURE, "", lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+    for (auto this_file : files) {
+      auto destination = fs::path(this_file.get<std::string>("destination"));
+      auto file_type = this_file.get<std::string>("file_type");
+      if (file_type == "file") {
+        try {
+          if (!fs::exists(destination.parent_path())) {
+            Util::createDir(destination.parent_path());
```

Comment:
Why is this necessary?

Reply:
Now that I think about it I think this is in the wrong place. downloadFileFromMaster doesn't create the directory itself, it downloads the file to a cachedir and renames it (using fs::rename()) if the SHA matches. So if the underlying destination dir for the file doesn't exist yet it will break.

Though after looking at it I think downloadFileFromMaster _should_ be creating the underlying destination dir if it doesn't exist, so that it can do so only after the matching SHA is determined.